Bluetooth: Host: bt_gatt_cb.att_mtu_updated only called when MTU exchange completes

ubieda opened the pull request:
This PR attempts to address #64172, by issuing the following changes:
1. `bt_gatt_cb.att_mtu_updated` is only called when the MTU exchange is completed (only once per connection).
2. Modified `tests/bsim/bluetooth/host/att/mtu_update` to directly verify MTU exchange takes place both on Central and Peripheral.
3. Updated Migration Guide document to reflect the behavior change of `att_mu_updated` as it was previously also triggered upon connection establishment (before MTU exchange).

NOTE: `att_mtu_updated` is not triggered if an MTU exchange results in an `ATT_ERROR_RSP` (e.g: Not supported). There's a comment on the issue whether that should also result in calling `att_mtu_updated`.


**Status on PR Actionable Items:**
- [ ] Modify triggering of `att_mtu_updated` callback.
    - [ ] For (U)ATT Bearers
        - [x]  Do not trigger on ATT connection.
        - [x] Trigger after successful MTU Exchange.
        - [ ] Trigger after a Failed MTU Exchange.
    - [x] For EATT Bearers
        - [x] May trigger on ATT connection, with updated MTU. 
        - [x] May trigger on Channel Reconfiguration.
- [ ] Capture test-cases to lock-in MTU update behavior.
    - [ ] For (U)ATT Bearers
        - [x] On Successfull MTU Exchange.
        - [ ] On Failed MTU exchange.
    - [ ] For EATT Bearers
        - [ ] On MTU-update from ATT connection.
        - [x] On MTU-update from ATT channel reconfiguration.

ubieda committed:
Bluetooth: Host: MTU Changed Notification only called after exchange

Per issue #64172, `att_mtu_updated` is expected to only notify once the
MTU exchange has taken place in a Bluetooth connection, and not once
connection is established. This allows each GATT profile to use this
callback as a way to be signaled when higher MTU is available, which is
expected to occur once per connection.
There is the exception for EATT L2CAP channels, where an MTU exchange
takes place at the L2CAP level, during ATT connection establishment.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
doc: releases: migration-guide: 3.6: noted att_mtu_updated change

Now is not triggered upon connection establishment.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
tests: bsim: bluetooth: host: att: Adjusted MTU test to verify exchange

Previously this sample verified the MTU exchange in an indirect manner
by being able to send a larger notification size. Now this adjustment
allows directly verifying both on the Central and the Peripheral that
the expected MTU exchange takes place, and that the listeners
subscribed through `bt_gatt_cb_register` are notified when this
exchange takes place (once per connection).

Signed-off-by: Luis Ubieda <luisf@croxel.com>

Thalley commented on samples/bluetooth/mtu_update/central/src/central_mtu_update.c:
```diff
@@ -205,7 +205,12 @@ static void connected(struct bt_conn *conn, uint8_t err)
 	printk("Connected: %s\n", addr);
 
-	(void)mtu_exchange(conn);
+	struct bt_conn_info info = {0};
```
Suggest to declare variables at top of scope (i.e. move this next to `addr` in line 192)

Thalley commented on samples/bluetooth/mtu_update/peripheral/src/main.c:
```diff
@@ -8,8 +8,17 @@
-extern void run_peripheral_sample(uint8_t *notify_data, size_t notify_data_size, uint16_t seconds);
+void mtu_updated(struct bt_conn *conn, uint16_t tx, uint16_t rx)
+{
+	printk("Updated MTU: TX: %d RX: %d bytes\n", tx, rx);
```
```suggestion
	printk("Updated MTU: TX: %u RX: %u bytes\n", tx, rx);
```

Thalley commented on samples/bluetooth/mtu_update/central/src/main.c:
```diff
@@ -10,10 +10,17 @@
-extern void run_central_sample(bt_gatt_notify_func_t cb);
+void mtu_updated(struct bt_conn *conn, uint16_t tx, uint16_t rx)
+{
+	printk("Updated MTU: TX: %d RX: %d bytes\n", tx, rx);
```
```suggestion
	printk("Updated MTU: TX: %u RX: %u bytes\n", tx, rx);
```

Thalley commented on samples/bluetooth/mtu_update/peripheral/src/peripheral_mtu_update.c:
```diff
@@ -56,7 +50,7 @@ void run_peripheral_sample(uint8_t *notify_data, size_t notify_data_size, uint16
 	}
 
-	bt_gatt_cb_register(&gatt_callbacks);
+	bt_gatt_cb_register(cb);
```
Similarly I would keep this sample's own callback and have the caller register it's own if it needs it.

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -3087,8 +3087,6 @@ static void bt_att_connected(struct bt_l2cap_chan *chan)
 
 	atomic_set_bit(att_chan->flags, ATT_CONNECTED);
 
-	att_chan_mtu_updated(att_chan);
-
```
I don't think this will work for EATT bearers, as the MTU is negotiated as par of the connection request, so if I understood correctly, this will not provide MTU information to the application on EATT connection established. We should probably extend one of the EATT babblesim tests to also care about the MTU exchange to ensure that the change we do here, works as expected for EATT as well.

Thalley commented on tests/bsim/bluetooth/host/att/mtu_update/src/main.c:
```diff
@@ -39,59 +41,84 @@ extern void run_peripheral_sample(uint8_t *notify_data, size_t notify_data_size,
-	printk("BSIM NOTIFY_CALLBACK\n");
-
-	is_data_equal = (length == sizeof(notify_data) && !memcmp(notify_data, data, length));
+#define CREATE_FLAG(flag) static atomic_t flag = 0
+#define INC_FLAG(flag) (void)atomic_inc(&flag)
+#define GET_FLAG(flag) atomic_get(&flag)
+#define WAIT_FOR_FLAG(flag)                                                                       \
+	while (atomic_get(&flag) == 0) {                                                          \
+		(void)k_sleep(K_MSEC(1));                                                         \
+	}
```
I understand why you've change this, but I'm unsure if they should keep the `FLAG` part of the name. Flags are typically either set of unset (boolean), and not an integer value

You could consider just making `central_mtu_count` a `volatile size_t` or a semaphore instead, and leave the remaining flags as is.

Thalley commented on tests/bsim/bluetooth/host/att/mtu_update/src/main.c:
```diff
@@ -39,59 +41,84 @@ extern void run_peripheral_sample(uint8_t *notify_data, size_t notify_data_size,
+		PASS("MTU Update test passed [Central]\n");
 	} else {
-		FAIL("MTU Update test failed\n");
+		FAIL("MTU Update test failed [Central] - Count: %d, MTU RX: %d, MTU TX: %d\n",
+				GET_FLAG(central_mtu_count), central_mtu_rx, central_mtu_tx);
```
```suggestion
		FAIL("MTU Update test failed [Central] - Count: %u, MTU RX: %u, MTU TX: %u\n",
				GET_FLAG(central_mtu_count), central_mtu_rx, central_mtu_tx);
```

Thalley replied:
Some indentation is also off here

Thalley commented on tests/bsim/bluetooth/host/att/mtu_update/src/main.c:
```diff
@@ -39,59 +41,84 @@ extern void run_peripheral_sample(uint8_t *notify_data, size_t notify_data_size,
+
+static void peripheral_mtu_updated(struct bt_conn *conn, uint16_t tx, uint16_t rx)
+{
+	printk("[PERIPHERAL] Updated MTU: Count: %d TX: %d RX: %d bytes\n",
+			peripheral_mtu_count, tx, rx);
```
```suggestion
	printk("[PERIPHERAL] Updated MTU: Count: %u TX: %u RX: %u bytes\n",
			peripheral_mtu_count, tx, rx);
```

Thalley replied:
Some indentation is also off here

Thalley commented on tests/bsim/bluetooth/host/att/mtu_update/src/main.c:
```diff
@@ -39,59 +41,84 @@ extern void run_peripheral_sample(uint8_t *notify_data, size_t notify_data_size,
 {
-	notify_data[13] = 0x7f;
-	notify_data[99] = 0x55;
+	uint16_t expected_mtu = 247;
```
```suggestion
	uint16_t expected_mtu = CONFIG_BT_L2CAP_TX_MTU;
```

Thalley commented on tests/bsim/bluetooth/host/att/mtu_update/src/main.c:
```diff
@@ -39,59 +41,84 @@ extern void run_peripheral_sample(uint8_t *notify_data, size_t notify_data_size,
 {
-	notify_data[13] = 0x7f;
-	notify_data[99] = 0x55;
+	uint16_t expected_mtu = 247;
```
```suggestion
	uint16_t expected_mtu = 247CONFIG_BT_L2CAP_TX_MTU
```

ubieda commented:
@Thalley thanks for your feedback!

I've addressed the following changes:
1. Restored APIs used for the mtu_exchange samples.
2. Added the `bt_gatt_cb_register()` before running the samples (both for peripheral and central).
3. Addressed formatting suggestions.

Follow-up Actions:
1. Make sure change works with EATT Bearers. Extend BabbleSIM test to do so.
> I don't think this will work for EATT bearers, as the MTU is negotiated as par of the connection request, so if I understood correctly, this will not provide MTU information to the application on EATT connection established. We should probably extend one of the EATT babblesim tests to also care about the MTU exchange to ensure that the change we do here, works as expected for EATT as well.
2. @Thalley Do you have any feedback on https://github.com/zephyrproject-rtos/zephyr/issues/64172#issuecomment-1899405393?

alwa-nordic commented:
I propose adding more test cases to nail down the behavior:
 - a test for MTU 23 case
 - a test that connects EATT but does not perform an ATT MTU exchange
 - a test for when the exchange results in an error

What do you think?

ubieda commented:
> I propose adding more test cases to nail down the behavior:
> 
> * a test for MTU 23 case
> * a test that connects EATT but does not perform an ATT MTU exchange
> * a test for when the exchange results in an error
> 
> What do you think?

@alwa-nordic Sounds good. I'll capture them in the tests!

Thalley commented on doc/releases/migration-guide-3.6.rst:
```diff
@@ -289,6 +289,9 @@ Bluetooth
 * The :c:func:`bt_l2cap_chan_send` API now requires the application to reserve
   enough bytes for the L2CAP headers. Call ``net_buf_reserve(buf,
   BT_L2CAP_SDU_CHAN_SEND_RESERVE);`` at buffer allocation time to do so.
+* The handler `att_mtu_updated` registered over the :c:func:`bt_gatt_cb_register`
+  API is no longer triggered upon connection establishment. It is now only
+  called when the MTU exchange has taken place.
```
It's still called on EATT connection establishment though

henrikbrixandersen replied:
This needs to go into the migration guide for v3.7.0 instead.

Thalley commented on samples/bluetooth/mtu_update/central/src/central_mtu_update.c:
```diff
@@ -205,7 +206,11 @@ static void connected(struct bt_conn *conn, uint8_t err)
-	(void)mtu_exchange(conn);
+	(void)bt_conn_get_info(conn, &info);
+
+	if (info.role == BT_CONN_ROLE_CENTRAL) {
+		(void)mtu_exchange(conn);
+	}
```
Seems superfluous to guard a function call with `BT_CONN_ROLE_CENTRAL` in a central sample.

Furthermore, both the central and peripheral is allowed to do the MTU exchange

Thalley commented on tests/bsim/bluetooth/host/att/mtu_update/src/main.c:
```diff
@@ -39,59 +39,85 @@ extern void run_peripheral_sample(uint8_t *notify_data, size_t notify_data_size,
+		k_sleep(K_MSEC(1));
+	}
+
+	if (central_mtu_count == 1 &&
+		central_mtu_tx == expected_mtu &&
+		central_mtu_rx == expected_mtu) {
+		PASS("MTU Update test passed [Central]\n");
```
Some indentation is off here

Thalley commented on tests/bsim/bluetooth/host/att/mtu_update/src/main.c:
```diff
@@ -39,59 +39,85 @@ extern void run_peripheral_sample(uint8_t *notify_data, size_t notify_data_size,
+		k_sleep(K_MSEC(1));
+	}
+
+	if (peripheral_mtu_count == 1 &&
+		peripheral_mtu_tx == expected_mtu &&
+		peripheral_mtu_rx == expected_mtu) {
```
Some indentation is off here

```suggestion
	if (peripheral_mtu_count == 1 &&
	    peripheral_mtu_tx == expected_mtu &&
	    peripheral_mtu_rx == expected_mtu) {
```

Thalley commented on tests/bsim/bluetooth/host/att/mtu_update/src/main.c:
```diff
@@ -39,59 +39,85 @@ extern void run_peripheral_sample(uint8_t *notify_data, size_t notify_data_size,
+		PASS("MTU Update test passed [Central]\n");
 	} else {
-		FAIL("MTU Update test failed\n");
+		FAIL("MTU Update test failed [Central] - Count: %u, MTU RX: %u, MTU TX: %u\n",
+			central_mtu_count, central_mtu_rx, central_mtu_tx);
```
```suggestion
		FAIL("MTU Update test failed [Central] - Count: %u, MTU RX: %u, MTU TX: %u\n",
		     central_mtu_count, central_mtu_rx, central_mtu_tx);
```

Thalley commented on tests/bsim/bluetooth/host/att/mtu_update/src/main.c:
```diff
@@ -39,59 +39,85 @@ extern void run_peripheral_sample(uint8_t *notify_data, size_t notify_data_size,
+		peripheral_mtu_rx == expected_mtu) {
+		PASS("MTU Update test passed [Peripheral]\n");
+	} else {
+		FAIL("MTU Update test failed [Peripheral] - Count: %u, MTU RX: %u, MTU TX: %u\n",
+			peripheral_mtu_count, peripheral_mtu_rx, peripheral_mtu_tx);
```
```suggestion
		FAIL("MTU Update test failed [Peripheral] - Count: %u, MTU RX: %u, MTU TX: %u\n",
		     peripheral_mtu_count, peripheral_mtu_rx, peripheral_mtu_tx);
```

Thalley commented on tests/bsim/bluetooth/host/att/mtu_update/src/main.c:
```diff
@@ -39,59 +39,85 @@ extern void run_peripheral_sample(uint8_t *notify_data, size_t notify_data_size,
+#define PERIPHERAL_WAIT_TIME ((WAIT_TIME - 10e6) / 1e6)
 
-CREATE_FLAG(flag_notification_received);
-uint8_t notify_data[100] = {};
-uint8_t is_data_equal;
```
Does it make sense to remove the notification part of the test? 

Wouldn't it make sense to both verify the MTU values **and** verify the notification length?

jori-nordic commented:
@ubieda sorry for the super-short notice, but do you want to join the meeting today (in an hour or so) to discuss this PR?

ubieda commented:
> @ubieda sorry for the super-short notice, but do you want to join the meeting today (in an hour or so) to discuss this PR?
> 

@jori-nordic Yeah - I'll see y'all in a few minutes!

ubieda commented:
Just to capture current status on this PR: It's marked as blocked per #68107. Once that is addressed, we can re-evaluate this PR and follow-up with the change-requests.

ubieda commented:
@Thalley @jori-nordic feel free to close if, after the latest changes in the BT Stack, this work needs to be restarted / no longer a suitable approach.

jori-nordic commented:
Yeah let's close it. It's linked to the RFC issue, so we can always go back and re-open it later. 
Thanks for putting in the work!
